pallets/nfts/src/features/approvals.rs

```diff
@@ -165,10 +184,207 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
+		collection: T::CollectionId,
+		witness_approvals: u32,
+	) -> DispatchResult {
+		let mut removed_approvals: u32 = 0;
+		let mut deposits: BalanceOf<T, I> = Zero::zero();
+		// Iterate and remove each collection approval, returning the deposited funds back to the
+		// `origin`.
+		for (_, (_, deposit)) in CollectionApprovals::<T, I>::drain_prefix((collection, &origin)) {
+			ensure!(removed_approvals < witness_approvals, Error::<T, I>::BadWitness);
+			removed_approvals.saturating_inc();
+			deposits = deposits.saturating_add(deposit);
+		}
+		ensure!(
+			removed_approvals > 0 && removed_approvals == witness_approvals,
+			Error::<T, I>::BadWitness
+		);
+		T::Currency::unreserve(&origin, deposits);
+		Self::deposit_event(Event::AllApprovalsCancelled { collection, item: None, owner: origin });
+		Ok(())
+	}
+
+	/// Checks whether the `delegate` is approved to transfer items in the `collection` owned
+	/// by the `account`.
+	///
+	/// - `collection`: The identifier of the collection.
+	/// - `account`: The account that granted the permission for `delegate` to transfer items.
+	/// - `delegate`: The account that had permission to transfer collection items owned by the
+	///   `account`.
+	fn check_collection_approval(
+		collection: &T::CollectionId,
+		account: &T::AccountId,
+		delegate: &T::AccountId,
+	) -> DispatchResult {
+		let (maybe_deadline, _) =
+			CollectionApprovals::<T, I>::get((&collection, &account, &delegate))
+				.ok_or(Error::<T, I>::NoPermission)?;
+		if let Some(deadline) = maybe_deadline {
+			let block_number = frame_system::Pallet::<T>::block_number();
+			ensure!(block_number <= deadline, Error::<T, I>::ApprovalExpired);
+		}
+		Ok(())
+	}
+
+	/// Checks whether the `delegate` is approved by the `account` to transfer items owned by
+	/// the `account` or a specific item in the collection. If the `delegate` has
+	/// an approval to transfer items in the collection owned by the `account`, they can
+	/// transfer every item without requiring explicit approval for that item.
+	///
+	/// - `collection`: The identifier of the collection.
+	/// - `maybe_item`: The optional item of the collection that the delegated account has an
+	///   approval to transfer. If not provided, an approval to transfer collection items owned by
+	///   the `account` will be checked.
+	/// - `account`: The account that granted the permission for `delegate` to transfer items in the
+	///   `collection` or the owner of the specified collection item.
+	/// - `delegate`: The account that had permission to transfer collection items owned by the
+	///   `account` or the specified collection item.
+	pub fn check_approval(
+		collection: &T::CollectionId,
+		maybe_item: &Option<T::ItemId>,
+		account: &T::AccountId,
+		delegate: &T::AccountId,
+	) -> DispatchResult {
+		// Check if a `delegate` has permission to transfer items in the collection owned by
+		// the `owner`.
+		match Self::check_collection_approval(collection, account, delegate) {
```

Comment:
- If `Item = None`

| Collection Approval  | Item Approval | Status |
| ------------- | ------------- | ------- |
| True  | False  | True |
| True  | True  | True |
| False | True | False |
| False  | False  | False |

- If `Item = Some`

| Collection Approval  | Item Approval | Status |
| ------------- | ------------- | ------- |
| True  | False  | True |
| True  | True  | True |
| False | True | True |
| False  | False  | False |

Reply:
This would be great to add somewhere when this is all done.

Reply:
Resolved! Already added to the PR description